File: std01-java-SwingApp/DecoratorPractice.java
/**
INSTRUCTIONS: 
    Demonstrates the Decorator design pattern by modeling a coffee beverage
    system where you can dynamically add ingredients (decorators) to a base beverage.

COMPILE & EXECUTE & CLEANUP (Java):

     javac  -d out              DecoratorPractice*.java
     java           -cp out     DecoratorPractice
     rm -rf out/
     












 */



public class DecoratorPractice {
    public static void main(String[] args) {
        print("Hello World");
    }

public interface Pizza {
    String Description();
    double cost();
}

public class NormalPizza implements Pizza {
    public String Description() {
        return "Pizza";
    }

    public double cost() {
        return 1.00;
    }
}

public abstract class Decorator implements Pizza {
    protected Pizza pizza;

    public Decorator(Pizza pizza) {
        this.pizza = pizza;
    }

    // public string Decorator() {
    //     return pizza.Decorator();
    // }

    public double cost() {
        return pizza.cost();
    }
}

















    public static void print(String text) {
        System.out.println(text);
    }
}

File: std01-java-SwingApp/DecoratorPractice.kt
/**
INSTRUCTIONS: 
    Running two different main methods (DecoratorPracticeKt.main / ClassShell.main) in Kotlin
    The 'Kt' suffix is only for `top-level functions`, not for `class-based main methods`.



COMPILE & EXECUTE & CLEANUP (Kotlin #1 - top-level main function):

     kotlinc    -d out              DecoratorPractice*.kt 
     kotlin             -cp out     DecoratorPracticeKt
     rm -rf out/

COMPILE & EXECUTE & CLEANUP (Kotlin #2 - class-based main method):

     kotlinc    -d out              DecoratorPractice*.kt
     kotlin             -cp out     ClassShell
     rm -rf out/




     

     
 */
fun main() {
    print("Hello World")
}

interface Pizza {
    fun description(): String
    fun cost(): Double
}

class NormalPizza : Pizza {
    override fun description(): String = "Pizza"
    override fun cost(): Double = 1.00
}

abstract class Decorator(protected val pizza: Pizza) : Pizza {
    override fun description(): String = pizza.description()
    override fun cost(): Double = pizza.cost()
}
